recipes/vnp46a2-h11v07/recipe.py

```diff
@@ -0,0 +1,139 @@
+"""
+A recipe to move VNP46A2 to a cloud analysis ready format.
+Based off briannapagan's GPM IMERG recipe.
+"""
+
+import datetime
+
+import pandas as pd
+import xarray as xr
+from cmr import GranuleQuery
+
+from pangeo_forge_recipes.patterns import ConcatDim, FilePattern
+from pangeo_forge_recipes.recipes import XarrayZarrRecipe
+
+# Query VNP46A2 dataset files at h11v07
+api = GranuleQuery()
+vnp_api = api.short_name('VNP46A2').point(-66, 18)  # Cerca de Puerto Rico (h11v07)
+granules = vnp_api.get_all()
+
+# Extract the link corresponding to each file
+downloadable_urls = []
+for g in granules:
+    for link in g['links']:
+
+        if link['rel'] == 'http://esipfed.org/ns/fedsearch/1.1/data#':
+            # print('adding url: ' + link['href'])
+            if not ('h11v07' in link['href']):
+                continue  # Double-checking we only capture h11v07
+            downloadable_urls.append(link['href'])
+            break
+
+print(f'{len(downloadable_urls)} urls')
+
+
+# Dictionaries containing the same information about each granule,
+# they just vary what variable you use as key to access them.
+vnp_date_dict = dict()  # Use granule date as key
+href_date_dict = dict()  # Granule download link as key
+
+vnp_dates = []  # List of granule dates, which is passed to ConcatDim later on
+
+for i in range(len(downloadable_urls)):
+
+    # Update broken links from the CMR for this dataset
+    href_orig = downloadable_urls[i]
+    href_new = (
+        href_orig.replace(
+            'https://ladsweb.modaps.eosdis.nasa.gov/archive/',
+            'https://ladsweb.modaps.eosdis.nasa.gov/opendap/RemoteResources/laads/',
+        )
+        + '.nc4'
+    )
+
+    # Convert julian date string to Python date object
+    year_julian = '-'.join(href_new.split('/')[-3:-1])
+    date_href = datetime.datetime.strptime(year_julian, '%Y-%j').date()
+
+    # Save this info into each dictionary and list
+    info_dict = dict(
+        i=i,
+        href=href_new,
+        date=date_href,
+        year_julian=year_julian,
+    )
+    vnp_date_dict[date_href] = info_dict
+    href_date_dict[href_new] = info_dict
+    vnp_dates.append(date_href)
+
+print('Earliest date:', min(vnp_dates).strftime('%Y-%m-%d'))
+print('Latest date:  ', max(vnp_dates).strftime('%Y-%m-%d'))
+
+
+def make_full_path(date: datetime.date) -> str:
+    '''
+    For each date, return the URL from the collected dictionary.
+    '''
+    return vnp_date_dict[date]['href']
```

Comment:
@jzavala-gonzalez, the latest test run failed due to a `NameError`. This isn't necessary a Python problem but one of the quirks of apache-beam: accessing variables defined in the global scope from a function scope seems to be a problem for apache-beam. one solution is to find a way to move `vnp_date_dict` from the global scope into `make_full_path` function scope by 

- defining `vnp_date_dict` inside `make_full_path` or
- passing `vpn_date_dict` as an argument 


```python-traceback
    File "/usr/local/lib/python3.9/dist-packages/pangeo_forge_recipes/executors/beam.py", line 40, in exec_stage
    File "/srv/conda/envs/notebook/lib/python3.9/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py", line 155, in cache_input
      fname = config.file_pattern[input_key]
    File "/srv/conda/envs/notebook/lib/python3.9/site-packages/pangeo_forge_recipes/patterns.py", line 219, in __getitem__
      fname = self.format_function(**format_function_kwargs)
    File "/tmp/tmpxgkkg7a3/recipes/vnp46a2-h11v07/recipe.py", line 77, in make_full_path
  NameError: name 'vnp_date_dict' is not defined [while running 'Start|cache_input|Reshuffle_000|prepare_target|Reshuffle_001|store_chunk|Reshuffle_002|finalize_target|Reshuffle_003/cache_input/Execute-ptransform-56']
```

Reply:
if you decide to go with option 2), i'm curious as to whether the following would work 


```python

import functools 

.... 

print('Earliest date:', min(vnp_dates).strftime('%Y-%m-%d'))
print('Latest date:  ', max(vnp_dates).strftime('%Y-%m-%d'))


def make_full_path(date: datetime.date, vnp_date_dict=None) -> str:
   return vnp_date_dict[date]['href']

... 

make_full_path = functoos.partial(make_full_path, vnp_date_dict=vnp_date_dict)

....


```

Reply:
Option 2 looks to be working for the pattern function as you wrote. I suspected the same would happen for the process_input function, `add_date_dimension`, but during the init of the XarrayZarrRecipe it's attempting unsuccesfully to JSON serialize that partial function (`TypeError: object of type partial not serializable`). In that second function it's much easier to adjust to not call global so I applied option 1 for that instead.